// @devvit/protos/types/devvit/runtime/actor/v1alpha/health.twirp.js
import { TwirpServer, TwirpError, TwirpErrorCode, TwirpContentType, chainInterceptors, } from 'twirp-ts';
import { Empty } from '../../../../google/protobuf/empty.js';
import { HealthStatus } from './health.js';
export var HealthMethod;
(function (HealthMethod) {
    HealthMethod["Check"] = "Check";
})(HealthMethod || (HealthMethod = {}));
export const HealthMethodList = [HealthMethod.Check];
export function createHealthServer(service) {
    return new TwirpServer({
        service,
        packageName: 'devvit.runtime.actor.v1alpha',
        serviceName: 'Health',
        methodList: HealthMethodList,
        matchRoute: matchHealthRoute,
    });
}
function matchHealthRoute(method, events) {
    switch (method) {
        case 'Check':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'Check' };
                await events.onMatch(ctx);
                return handleHealthCheckRequest(ctx, service, data, interceptors);
            };
        default:
            events.onNotFound();
            const msg = `no handler found`;
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleHealthCheckRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleHealthCheckJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleHealthCheckProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
async function handleHealthCheckJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = Empty.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Check(ctx, inputReq);
        });
    }
    else {
        response = await service.Check(ctx, request);
    }
    return JSON.stringify(HealthStatus.toJSON(response));
}
async function handleHealthCheckProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = Empty.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Check(ctx, inputReq);
        });
    }
    else {
        response = await service.Check(ctx, request);
    }
    return Buffer.from(HealthStatus.encode(response).finish());
}
