// @devvit/protos/types/devvit/debug/v1alpha/debug.twirp.js
import { TwirpServer, TwirpError, TwirpErrorCode, TwirpContentType, chainInterceptors, } from 'twirp-ts';
import { GetProfileRequest, GetProfileResponse } from './debug.js';
export var DevvitProfilerMethod;
(function (DevvitProfilerMethod) {
    DevvitProfilerMethod["GetProfile"] = "GetProfile";
})(DevvitProfilerMethod || (DevvitProfilerMethod = {}));
export const DevvitProfilerMethodList = [DevvitProfilerMethod.GetProfile];
export function createDevvitProfilerServer(service) {
    return new TwirpServer({
        service,
        packageName: 'devvit.debug.v1alpha',
        serviceName: 'DevvitProfiler',
        methodList: DevvitProfilerMethodList,
        matchRoute: matchDevvitProfilerRoute,
    });
}
function matchDevvitProfilerRoute(method, events) {
    switch (method) {
        case 'GetProfile':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'GetProfile' };
                await events.onMatch(ctx);
                return handleDevvitProfilerGetProfileRequest(ctx, service, data, interceptors);
            };
        default:
            events.onNotFound();
            const msg = `no handler found`;
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleDevvitProfilerGetProfileRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleDevvitProfilerGetProfileJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleDevvitProfilerGetProfileProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
async function handleDevvitProfilerGetProfileJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = GetProfileRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.GetProfile(ctx, inputReq);
        });
    }
    else {
        response = await service.GetProfile(ctx, request);
    }
    return JSON.stringify(GetProfileResponse.toJSON(response));
}
async function handleDevvitProfilerGetProfileProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = GetProfileRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.GetProfile(ctx, inputReq);
        });
    }
    else {
        response = await service.GetProfile(ctx, request);
    }
    return Buffer.from(GetProfileResponse.encode(response).finish());
}
