// @devvit/protos/types/devvit/plugin/clock/clock.twirp.js
import { TwirpServer, TwirpError, TwirpErrorCode, TwirpContentType, chainInterceptors, } from 'twirp-ts';
import { Empty } from '../../../google/protobuf/empty.js';
import { Timestamp } from '../../../google/protobuf/timestamp.js';
export var ClockMethod;
(function (ClockMethod) {
    ClockMethod["Now"] = "Now";
})(ClockMethod || (ClockMethod = {}));
export const ClockMethodList = [ClockMethod.Now];
export function createClockServer(service) {
    return new TwirpServer({
        service,
        packageName: 'devvit.plugin.clock',
        serviceName: 'Clock',
        methodList: ClockMethodList,
        matchRoute: matchClockRoute,
    });
}
function matchClockRoute(method, events) {
    switch (method) {
        case 'Now':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'Now' };
                await events.onMatch(ctx);
                return handleClockNowRequest(ctx, service, data, interceptors);
            };
        default:
            events.onNotFound();
            const msg = `no handler found`;
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleClockNowRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleClockNowJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleClockNowProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
async function handleClockNowJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = Empty.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Now(ctx, inputReq);
        });
    }
    else {
        response = await service.Now(ctx, request);
    }
    return JSON.stringify(Timestamp.toJSON(response));
}
async function handleClockNowProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = Empty.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Now(ctx, inputReq);
        });
    }
    else {
        response = await service.Now(ctx, request);
    }
    return Buffer.from(Timestamp.encode(response).finish());
}
