// @devvit/protos/types/devvit/plugin/timer/timer.twirp.js
import { TwirpServer, TwirpError, TwirpErrorCode, TwirpContentType, chainInterceptors, } from 'twirp-ts';
import { TimerMessage } from './timer.js';
import { Empty } from '../../../google/protobuf/empty.js';
export var TimerMethod;
(function (TimerMethod) {
    TimerMethod["Wait"] = "Wait";
})(TimerMethod || (TimerMethod = {}));
export const TimerMethodList = [TimerMethod.Wait];
export function createTimerServer(service) {
    return new TwirpServer({
        service,
        packageName: 'devvit.plugin.timer',
        serviceName: 'Timer',
        methodList: TimerMethodList,
        matchRoute: matchTimerRoute,
    });
}
function matchTimerRoute(method, events) {
    switch (method) {
        case 'Wait':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'Wait' };
                await events.onMatch(ctx);
                return handleTimerWaitRequest(ctx, service, data, interceptors);
            };
        default:
            events.onNotFound();
            const msg = `no handler found`;
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleTimerWaitRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleTimerWaitJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleTimerWaitProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
async function handleTimerWaitJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = TimerMessage.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Wait(ctx, inputReq);
        });
    }
    else {
        response = await service.Wait(ctx, request);
    }
    return JSON.stringify(Empty.toJSON(response));
}
async function handleTimerWaitProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = TimerMessage.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Wait(ctx, inputReq);
        });
    }
    else {
        response = await service.Wait(ctx, request);
    }
    return Buffer.from(Empty.encode(response).finish());
}
