// @devvit/protos/types/devvit/runtime/actor/echo.twirp.js
import { TwirpServer, TwirpError, TwirpErrorCode, TwirpContentType, chainInterceptors, } from 'twirp-ts';
import { StringValue } from '../../../google/protobuf/wrappers.js';
export var EchoMethod;
(function (EchoMethod) {
    EchoMethod["Echo"] = "Echo";
})(EchoMethod || (EchoMethod = {}));
export const EchoMethodList = [EchoMethod.Echo];
export function createEchoServer(service) {
    return new TwirpServer({
        service,
        packageName: 'devvit.runtime.actor',
        serviceName: 'Echo',
        methodList: EchoMethodList,
        matchRoute: matchEchoRoute,
    });
}
function matchEchoRoute(method, events) {
    switch (method) {
        case 'Echo':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'Echo' };
                await events.onMatch(ctx);
                return handleEchoEchoRequest(ctx, service, data, interceptors);
            };
        default:
            events.onNotFound();
            const msg = `no handler found`;
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleEchoEchoRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleEchoEchoJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleEchoEchoProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
async function handleEchoEchoJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = StringValue.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Echo(ctx, inputReq);
        });
    }
    else {
        response = await service.Echo(ctx, request);
    }
    return JSON.stringify(StringValue.toJSON(response));
}
async function handleEchoEchoProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = StringValue.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Echo(ctx, inputReq);
        });
    }
    else {
        response = await service.Echo(ctx, request);
    }
    return Buffer.from(StringValue.encode(response).finish());
}
