// @devvit/protos/types/devvit/actor/settings/v1alpha/app_settings.twirp.js
import { TwirpServer, TwirpError, TwirpErrorCode, TwirpContentType, chainInterceptors, } from 'twirp-ts';
import { GetFieldsRequest, GetFieldsResponse, ValidateFormRequest, ValidateFormResponse, } from './shared.js';
export var AppSettingsMethod;
(function (AppSettingsMethod) {
    AppSettingsMethod["GetAppSettingsFields"] = "GetAppSettingsFields";
    AppSettingsMethod["ValidateAppForm"] = "ValidateAppForm";
})(AppSettingsMethod || (AppSettingsMethod = {}));
export const AppSettingsMethodList = [
    AppSettingsMethod.GetAppSettingsFields,
    AppSettingsMethod.ValidateAppForm,
];
export function createAppSettingsServer(service) {
    return new TwirpServer({
        service,
        packageName: 'devvit.actor.settings.v1alpha',
        serviceName: 'AppSettings',
        methodList: AppSettingsMethodList,
        matchRoute: matchAppSettingsRoute,
    });
}
function matchAppSettingsRoute(method, events) {
    switch (method) {
        case 'GetAppSettingsFields':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'GetAppSettingsFields' };
                await events.onMatch(ctx);
                return handleAppSettingsGetAppSettingsFieldsRequest(ctx, service, data, interceptors);
            };
        case 'ValidateAppForm':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'ValidateAppForm' };
                await events.onMatch(ctx);
                return handleAppSettingsValidateAppFormRequest(ctx, service, data, interceptors);
            };
        default:
            events.onNotFound();
            const msg = `no handler found`;
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleAppSettingsGetAppSettingsFieldsRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleAppSettingsGetAppSettingsFieldsJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleAppSettingsGetAppSettingsFieldsProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleAppSettingsValidateAppFormRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleAppSettingsValidateAppFormJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleAppSettingsValidateAppFormProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
async function handleAppSettingsGetAppSettingsFieldsJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = GetFieldsRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.GetAppSettingsFields(ctx, inputReq);
        });
    }
    else {
        response = await service.GetAppSettingsFields(ctx, request);
    }
    return JSON.stringify(GetFieldsResponse.toJSON(response));
}
async function handleAppSettingsValidateAppFormJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = ValidateFormRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.ValidateAppForm(ctx, inputReq);
        });
    }
    else {
        response = await service.ValidateAppForm(ctx, request);
    }
    return JSON.stringify(ValidateFormResponse.toJSON(response));
}
async function handleAppSettingsGetAppSettingsFieldsProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = GetFieldsRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.GetAppSettingsFields(ctx, inputReq);
        });
    }
    else {
        response = await service.GetAppSettingsFields(ctx, request);
    }
    return Buffer.from(GetFieldsResponse.encode(response).finish());
}
async function handleAppSettingsValidateAppFormProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = ValidateFormRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.ValidateAppForm(ctx, inputReq);
        });
    }
    else {
        response = await service.ValidateAppForm(ctx, request);
    }
    return Buffer.from(ValidateFormResponse.encode(response).finish());
}
