// @devvit/protos/types/devvit/actor/settings/v1alpha/installation_settings.twirp.js
import { TwirpServer, TwirpError, TwirpErrorCode, TwirpContentType, chainInterceptors, } from 'twirp-ts';
import { GetFieldsRequest, GetFieldsResponse, ValidateFormRequest, ValidateFormResponse, } from './shared.js';
export var InstallationSettingsMethod;
(function (InstallationSettingsMethod) {
    InstallationSettingsMethod["GetSettingsFields"] = "GetSettingsFields";
    InstallationSettingsMethod["ValidateForm"] = "ValidateForm";
})(InstallationSettingsMethod || (InstallationSettingsMethod = {}));
export const InstallationSettingsMethodList = [
    InstallationSettingsMethod.GetSettingsFields,
    InstallationSettingsMethod.ValidateForm,
];
export function createInstallationSettingsServer(service) {
    return new TwirpServer({
        service,
        packageName: 'devvit.actor.settings.v1alpha',
        serviceName: 'InstallationSettings',
        methodList: InstallationSettingsMethodList,
        matchRoute: matchInstallationSettingsRoute,
    });
}
function matchInstallationSettingsRoute(method, events) {
    switch (method) {
        case 'GetSettingsFields':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'GetSettingsFields' };
                await events.onMatch(ctx);
                return handleInstallationSettingsGetSettingsFieldsRequest(ctx, service, data, interceptors);
            };
        case 'ValidateForm':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'ValidateForm' };
                await events.onMatch(ctx);
                return handleInstallationSettingsValidateFormRequest(ctx, service, data, interceptors);
            };
        default:
            events.onNotFound();
            const msg = `no handler found`;
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleInstallationSettingsGetSettingsFieldsRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleInstallationSettingsGetSettingsFieldsJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleInstallationSettingsGetSettingsFieldsProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleInstallationSettingsValidateFormRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleInstallationSettingsValidateFormJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleInstallationSettingsValidateFormProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
async function handleInstallationSettingsGetSettingsFieldsJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = GetFieldsRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.GetSettingsFields(ctx, inputReq);
        });
    }
    else {
        response = await service.GetSettingsFields(ctx, request);
    }
    return JSON.stringify(GetFieldsResponse.toJSON(response));
}
async function handleInstallationSettingsValidateFormJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = ValidateFormRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.ValidateForm(ctx, inputReq);
        });
    }
    else {
        response = await service.ValidateForm(ctx, request);
    }
    return JSON.stringify(ValidateFormResponse.toJSON(response));
}
async function handleInstallationSettingsGetSettingsFieldsProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = GetFieldsRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.GetSettingsFields(ctx, inputReq);
        });
    }
    else {
        response = await service.GetSettingsFields(ctx, request);
    }
    return Buffer.from(GetFieldsResponse.encode(response).finish());
}
async function handleInstallationSettingsValidateFormProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = ValidateFormRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.ValidateForm(ctx, inputReq);
        });
    }
    else {
        response = await service.ValidateForm(ctx, request);
    }
    return Buffer.from(ValidateFormResponse.encode(response).finish());
}
