// @devvit/protos/types/devvit/service/envelope_service.twirp.js
import { TwirpServer, TwirpError, TwirpErrorCode, TwirpContentType, chainInterceptors, } from 'twirp-ts';
import { Envelope } from '../runtime/envelope.js';
import { Empty } from '../../google/protobuf/empty.js';
export var EnvelopeReceiverPortMethod;
(function (EnvelopeReceiverPortMethod) {
    EnvelopeReceiverPortMethod["Post"] = "Post";
})(EnvelopeReceiverPortMethod || (EnvelopeReceiverPortMethod = {}));
export const EnvelopeReceiverPortMethodList = [EnvelopeReceiverPortMethod.Post];
export function createEnvelopeReceiverPortServer(service) {
    return new TwirpServer({
        service,
        packageName: 'devvit.service',
        serviceName: 'EnvelopeReceiverPort',
        methodList: EnvelopeReceiverPortMethodList,
        matchRoute: matchEnvelopeReceiverPortRoute,
    });
}
function matchEnvelopeReceiverPortRoute(method, events) {
    switch (method) {
        case 'Post':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'Post' };
                await events.onMatch(ctx);
                return handleEnvelopeReceiverPortPostRequest(ctx, service, data, interceptors);
            };
        default:
            events.onNotFound();
            const msg = `no handler found`;
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleEnvelopeReceiverPortPostRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleEnvelopeReceiverPortPostJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleEnvelopeReceiverPortPostProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
async function handleEnvelopeReceiverPortPostJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = Envelope.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Post(ctx, inputReq);
        });
    }
    else {
        response = await service.Post(ctx, request);
    }
    return JSON.stringify(Empty.toJSON(response));
}
async function handleEnvelopeReceiverPortPostProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = Envelope.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Post(ctx, inputReq);
        });
    }
    else {
        response = await service.Post(ctx, request);
    }
    return Buffer.from(Empty.encode(response).finish());
}
