// @devvit/protos/types/devvit/ui/events/v1alpha/handle_ui.twirp.js
import { TwirpServer, TwirpError, TwirpErrorCode, TwirpContentType, chainInterceptors, } from 'twirp-ts';
import { HandleUIEventRequest, HandleUIEventResponse } from './handle_ui.js';
export var UIEventHandlerMethod;
(function (UIEventHandlerMethod) {
    UIEventHandlerMethod["HandleUIEvent"] = "HandleUIEvent";
})(UIEventHandlerMethod || (UIEventHandlerMethod = {}));
export const UIEventHandlerMethodList = [UIEventHandlerMethod.HandleUIEvent];
export function createUIEventHandlerServer(service) {
    return new TwirpServer({
        service,
        packageName: 'devvit.ui.events.v1alpha',
        serviceName: 'UIEventHandler',
        methodList: UIEventHandlerMethodList,
        matchRoute: matchUIEventHandlerRoute,
    });
}
function matchUIEventHandlerRoute(method, events) {
    switch (method) {
        case 'HandleUIEvent':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'HandleUIEvent' };
                await events.onMatch(ctx);
                return handleUIEventHandlerHandleUIEventRequest(ctx, service, data, interceptors);
            };
        default:
            events.onNotFound();
            const msg = `no handler found`;
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleUIEventHandlerHandleUIEventRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleUIEventHandlerHandleUIEventJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleUIEventHandlerHandleUIEventProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
async function handleUIEventHandlerHandleUIEventJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = HandleUIEventRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.HandleUIEvent(ctx, inputReq);
        });
    }
    else {
        response = await service.HandleUIEvent(ctx, request);
    }
    return JSON.stringify(HandleUIEventResponse.toJSON(response));
}
async function handleUIEventHandlerHandleUIEventProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = HandleUIEventRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.HandleUIEvent(ctx, inputReq);
        });
    }
    else {
        response = await service.HandleUIEvent(ctx, request);
    }
    return Buffer.from(HandleUIEventResponse.encode(response).finish());
}
