// @devvit/protos/types/devvit/plugin/media/media.twirp.js
import { TwirpServer, TwirpError, TwirpErrorCode, TwirpContentType, chainInterceptors, } from 'twirp-ts';
import { MediaUploadRequest, MediaUploadResponse } from './media.js';
export var MediaServiceMethod;
(function (MediaServiceMethod) {
    MediaServiceMethod["Upload"] = "Upload";
})(MediaServiceMethod || (MediaServiceMethod = {}));
export const MediaServiceMethodList = [MediaServiceMethod.Upload];
export function createMediaServiceServer(service) {
    return new TwirpServer({
        service,
        packageName: 'devvit.plugin.media',
        serviceName: 'MediaService',
        methodList: MediaServiceMethodList,
        matchRoute: matchMediaServiceRoute,
    });
}
function matchMediaServiceRoute(method, events) {
    switch (method) {
        case 'Upload':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'Upload' };
                await events.onMatch(ctx);
                return handleMediaServiceUploadRequest(ctx, service, data, interceptors);
            };
        default:
            events.onNotFound();
            const msg = `no handler found`;
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleMediaServiceUploadRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleMediaServiceUploadJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleMediaServiceUploadProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
async function handleMediaServiceUploadJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = MediaUploadRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Upload(ctx, inputReq);
        });
    }
    else {
        response = await service.Upload(ctx, request);
    }
    return JSON.stringify(MediaUploadResponse.toJSON(response));
}
async function handleMediaServiceUploadProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = MediaUploadRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Upload(ctx, inputReq);
        });
    }
    else {
        response = await service.Upload(ctx, request);
    }
    return Buffer.from(MediaUploadResponse.encode(response).finish());
}
