// @devvit/protos/types/devvit/actor/producer/producer.twirp.js
import { TwirpServer, TwirpError, TwirpErrorCode, TwirpContentType, chainInterceptors, } from 'twirp-ts';
import { Args, Arg } from './producer.js';
export var ProducerMethod;
(function (ProducerMethod) {
    ProducerMethod["Produce"] = "Produce";
})(ProducerMethod || (ProducerMethod = {}));
export const ProducerMethodList = [ProducerMethod.Produce];
export function createProducerServer(service) {
    return new TwirpServer({
        service,
        packageName: 'devvit.actor.producer',
        serviceName: 'Producer',
        methodList: ProducerMethodList,
        matchRoute: matchProducerRoute,
    });
}
function matchProducerRoute(method, events) {
    switch (method) {
        case 'Produce':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'Produce' };
                await events.onMatch(ctx);
                return handleProducerProduceRequest(ctx, service, data, interceptors);
            };
        default:
            events.onNotFound();
            const msg = `no handler found`;
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleProducerProduceRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleProducerProduceJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleProducerProduceProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
async function handleProducerProduceJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = Args.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Produce(ctx, inputReq);
        });
    }
    else {
        response = await service.Produce(ctx, request);
    }
    return JSON.stringify(Arg.toJSON(response));
}
async function handleProducerProduceProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = Args.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Produce(ctx, inputReq);
        });
    }
    else {
        response = await service.Produce(ctx, request);
    }
    return Buffer.from(Arg.encode(response).finish());
}
