// @devvit/protos/types/devvit/plugin/modlog/modlog.twirp.js
import { TwirpServer, TwirpError, TwirpErrorCode, TwirpContentType, chainInterceptors, } from 'twirp-ts';
import { ModlogRequest, ModlogResponse } from './modlog.js';
export var ModlogMethod;
(function (ModlogMethod) {
    ModlogMethod["Add"] = "Add";
})(ModlogMethod || (ModlogMethod = {}));
export const ModlogMethodList = [ModlogMethod.Add];
export function createModlogServer(service) {
    return new TwirpServer({
        service,
        packageName: 'devvit.plugin.modlog',
        serviceName: 'Modlog',
        methodList: ModlogMethodList,
        matchRoute: matchModlogRoute,
    });
}
function matchModlogRoute(method, events) {
    switch (method) {
        case 'Add':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'Add' };
                await events.onMatch(ctx);
                return handleModlogAddRequest(ctx, service, data, interceptors);
            };
        default:
            events.onNotFound();
            const msg = `no handler found`;
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleModlogAddRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleModlogAddJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleModlogAddProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
async function handleModlogAddJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = ModlogRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Add(ctx, inputReq);
        });
    }
    else {
        response = await service.Add(ctx, request);
    }
    return JSON.stringify(ModlogResponse.toJSON(response));
}
async function handleModlogAddProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = ModlogRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Add(ctx, inputReq);
        });
    }
    else {
        response = await service.Add(ctx, request);
    }
    return Buffer.from(ModlogResponse.encode(response).finish());
}
