// @devvit/protos/types/devvit/plugin/linker/linker.twirp.js
import { TwirpServer, TwirpError, TwirpErrorCode, TwirpContentType, chainInterceptors, } from 'twirp-ts';
import { Bundle } from '../buildpack/buildpack_common.js';
import { LinkerResponse } from './linker.js';
export var LinkerMethod;
(function (LinkerMethod) {
    LinkerMethod["Link"] = "Link";
})(LinkerMethod || (LinkerMethod = {}));
export const LinkerMethodList = [LinkerMethod.Link];
export function createLinkerServer(service) {
    return new TwirpServer({
        service,
        packageName: 'devvit.plugin.linker',
        serviceName: 'Linker',
        methodList: LinkerMethodList,
        matchRoute: matchLinkerRoute,
    });
}
function matchLinkerRoute(method, events) {
    switch (method) {
        case 'Link':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'Link' };
                await events.onMatch(ctx);
                return handleLinkerLinkRequest(ctx, service, data, interceptors);
            };
        default:
            events.onNotFound();
            const msg = `no handler found`;
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleLinkerLinkRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleLinkerLinkJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleLinkerLinkProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
async function handleLinkerLinkJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = Bundle.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Link(ctx, inputReq);
        });
    }
    else {
        response = await service.Link(ctx, request);
    }
    return JSON.stringify(LinkerResponse.toJSON(response));
}
async function handleLinkerLinkProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = Bundle.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Link(ctx, inputReq);
        });
    }
    else {
        response = await service.Link(ctx, request);
    }
    return Buffer.from(LinkerResponse.encode(response).finish());
}
