// @devvit/protos/types/devvit/plugin/http/http.twirp.js
import { TwirpServer, TwirpError, TwirpErrorCode, TwirpContentType, chainInterceptors, } from 'twirp-ts';
import { FetchRequest, FetchResponse } from './http.js';
export var HTTPMethod;
(function (HTTPMethod) {
    HTTPMethod["Fetch"] = "Fetch";
})(HTTPMethod || (HTTPMethod = {}));
export const HTTPMethodList = [HTTPMethod.Fetch];
export function createHTTPServer(service) {
    return new TwirpServer({
        service,
        packageName: 'devvit.plugin.http',
        serviceName: 'HTTP',
        methodList: HTTPMethodList,
        matchRoute: matchHTTPRoute,
    });
}
function matchHTTPRoute(method, events) {
    switch (method) {
        case 'Fetch':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'Fetch' };
                await events.onMatch(ctx);
                return handleHTTPFetchRequest(ctx, service, data, interceptors);
            };
        default:
            events.onNotFound();
            const msg = `no handler found`;
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleHTTPFetchRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleHTTPFetchJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleHTTPFetchProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
async function handleHTTPFetchJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = FetchRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Fetch(ctx, inputReq);
        });
    }
    else {
        response = await service.Fetch(ctx, request);
    }
    return JSON.stringify(FetchResponse.toJSON(response));
}
async function handleHTTPFetchProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = FetchRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Fetch(ctx, inputReq);
        });
    }
    else {
        response = await service.Fetch(ctx, request);
    }
    return Buffer.from(FetchResponse.encode(response).finish());
}
