// @devvit/protos/types/devvit/gateway/resolvers.twirp.js
import { TwirpServer, TwirpError, TwirpErrorCode, TwirpContentType, chainInterceptors, } from 'twirp-ts';
import { SubredditMetadataRequest } from './resolvers.js';
import { DevvitSubredditMetadata } from '../gql/subreddit_metadata.js';
export var SubredditMetadataResolverMethod;
(function (SubredditMetadataResolverMethod) {
    SubredditMetadataResolverMethod["GetSubredditMetadata"] = "GetSubredditMetadata";
})(SubredditMetadataResolverMethod || (SubredditMetadataResolverMethod = {}));
export const SubredditMetadataResolverMethodList = [
    SubredditMetadataResolverMethod.GetSubredditMetadata,
];
export function createSubredditMetadataResolverServer(service) {
    return new TwirpServer({
        service,
        packageName: 'devvit.gateway',
        serviceName: 'SubredditMetadataResolver',
        methodList: SubredditMetadataResolverMethodList,
        matchRoute: matchSubredditMetadataResolverRoute,
    });
}
function matchSubredditMetadataResolverRoute(method, events) {
    switch (method) {
        case 'GetSubredditMetadata':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'GetSubredditMetadata' };
                await events.onMatch(ctx);
                return handleSubredditMetadataResolverGetSubredditMetadataRequest(ctx, service, data, interceptors);
            };
        default:
            events.onNotFound();
            const msg = `no handler found`;
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleSubredditMetadataResolverGetSubredditMetadataRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleSubredditMetadataResolverGetSubredditMetadataJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleSubredditMetadataResolverGetSubredditMetadataProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
async function handleSubredditMetadataResolverGetSubredditMetadataJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = SubredditMetadataRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.GetSubredditMetadata(ctx, inputReq);
        });
    }
    else {
        response = await service.GetSubredditMetadata(ctx, request);
    }
    return JSON.stringify(DevvitSubredditMetadata.toJSON(response));
}
async function handleSubredditMetadataResolverGetSubredditMetadataProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = SubredditMetadataRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.GetSubredditMetadata(ctx, inputReq);
        });
    }
    else {
        response = await service.GetSubredditMetadata(ctx, request);
    }
    return Buffer.from(DevvitSubredditMetadata.encode(response).finish());
}
