// @devvit/protos/types/devvit/dev_portal/app_publish_request/review/app_publish_request_review.js
/**
 * #app_publish_request_review.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
import _m0 from "protobufjs/minimal.js";
import { Timestamp } from "../../../../google/protobuf/timestamp.js";
import { messageTypeRegistry } from "../../../../typeRegistry.js";
import { Redditor } from "../../reddit/redditor.js";
export var AppPublishRequestReviewDecision;
(function (AppPublishRequestReviewDecision) {
    AppPublishRequestReviewDecision[AppPublishRequestReviewDecision["APPROVED"] = 0] = "APPROVED";
    AppPublishRequestReviewDecision[AppPublishRequestReviewDecision["DENIED"] = 1] = "DENIED";
    AppPublishRequestReviewDecision[AppPublishRequestReviewDecision["OBSOLETE"] = 2] = "OBSOLETE";
    AppPublishRequestReviewDecision[AppPublishRequestReviewDecision["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(AppPublishRequestReviewDecision || (AppPublishRequestReviewDecision = {}));
export function appPublishRequestReviewDecisionFromJSON(object) {
    switch (object) {
        case 0:
        case "APPROVED":
            return AppPublishRequestReviewDecision.APPROVED;
        case 1:
        case "DENIED":
            return AppPublishRequestReviewDecision.DENIED;
        case 2:
        case "OBSOLETE":
            return AppPublishRequestReviewDecision.OBSOLETE;
        case -1:
        case "UNRECOGNIZED":
        default:
            return AppPublishRequestReviewDecision.UNRECOGNIZED;
    }
}
export function appPublishRequestReviewDecisionToJSON(object) {
    switch (object) {
        case AppPublishRequestReviewDecision.APPROVED:
            return 0;
        case AppPublishRequestReviewDecision.DENIED:
            return 1;
        case AppPublishRequestReviewDecision.OBSOLETE:
            return 2;
        case AppPublishRequestReviewDecision.UNRECOGNIZED:
        default:
            return -1;
    }
}
function createBaseAppPublishRequestReview() {
    return { id: "", createdAt: undefined, updatedAt: undefined, reviewer: undefined, decision: 0 };
}
export const AppPublishRequestReview = {
    $type: "devvit.dev_portal.app_publish_request.review.AppPublishRequestReview",
    encode(message, writer = _m0.Writer.create()) {
        if (message.id !== "") {
            writer.uint32(10).string(message.id);
        }
        if (message.createdAt !== undefined) {
            Timestamp.encode(toTimestamp(message.createdAt), writer.uint32(18).fork()).ldelim();
        }
        if (message.updatedAt !== undefined) {
            Timestamp.encode(toTimestamp(message.updatedAt), writer.uint32(26).fork()).ldelim();
        }
        if (message.reviewer !== undefined) {
            Redditor.encode(message.reviewer, writer.uint32(34).fork()).ldelim();
        }
        if (message.decision !== 0) {
            writer.uint32(40).int32(message.decision);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAppPublishRequestReview();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.id = reader.string();
                    continue;
                case 2:
                    if (tag !== 18) {
                        break;
                    }
                    message.createdAt = fromTimestamp(Timestamp.decode(reader, reader.uint32()));
                    continue;
                case 3:
                    if (tag !== 26) {
                        break;
                    }
                    message.updatedAt = fromTimestamp(Timestamp.decode(reader, reader.uint32()));
                    continue;
                case 4:
                    if (tag !== 34) {
                        break;
                    }
                    message.reviewer = Redditor.decode(reader, reader.uint32());
                    continue;
                case 5:
                    if (tag !== 40) {
                        break;
                    }
                    message.decision = reader.int32();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            id: isSet(object.id) ? globalThis.String(object.id) : "",
            createdAt: isSet(object.createdAt) ? fromJsonTimestamp(object.createdAt) : undefined,
            updatedAt: isSet(object.updatedAt) ? fromJsonTimestamp(object.updatedAt) : undefined,
            reviewer: isSet(object.reviewer) ? Redditor.fromJSON(object.reviewer) : undefined,
            decision: isSet(object.decision) ? appPublishRequestReviewDecisionFromJSON(object.decision) : 0,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.id !== "") {
            obj.id = message.id;
        }
        if (message.createdAt !== undefined) {
            obj.createdAt = message.createdAt.toISOString();
        }
        if (message.updatedAt !== undefined) {
            obj.updatedAt = message.updatedAt.toISOString();
        }
        if (message.reviewer !== undefined) {
            obj.reviewer = Redditor.toJSON(message.reviewer);
        }
        if (message.decision !== 0) {
            obj.decision = appPublishRequestReviewDecisionToJSON(message.decision);
        }
        return obj;
    },
    create(base) {
        return AppPublishRequestReview.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseAppPublishRequestReview();
        message.id = object.id ?? "";
        message.createdAt = object.createdAt ?? undefined;
        message.updatedAt = object.updatedAt ?? undefined;
        message.reviewer = (object.reviewer !== undefined && object.reviewer !== null)
            ? Redditor.fromPartial(object.reviewer)
            : undefined;
        message.decision = object.decision ?? 0;
        return message;
    },
};
messageTypeRegistry.set(AppPublishRequestReview.$type, AppPublishRequestReview);
function createBaseAppPRSetReviewDecisionRequest() {
    return { appPublishRequestId: "", decision: 0 };
}
export const AppPRSetReviewDecisionRequest = {
    $type: "devvit.dev_portal.app_publish_request.review.AppPRSetReviewDecisionRequest",
    encode(message, writer = _m0.Writer.create()) {
        if (message.appPublishRequestId !== "") {
            writer.uint32(10).string(message.appPublishRequestId);
        }
        if (message.decision !== 0) {
            writer.uint32(16).int32(message.decision);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAppPRSetReviewDecisionRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.appPublishRequestId = reader.string();
                    continue;
                case 2:
                    if (tag !== 16) {
                        break;
                    }
                    message.decision = reader.int32();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            appPublishRequestId: isSet(object.appPublishRequestId) ? globalThis.String(object.appPublishRequestId) : "",
            decision: isSet(object.decision) ? appPublishRequestReviewDecisionFromJSON(object.decision) : 0,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.appPublishRequestId !== "") {
            obj.appPublishRequestId = message.appPublishRequestId;
        }
        if (message.decision !== 0) {
            obj.decision = appPublishRequestReviewDecisionToJSON(message.decision);
        }
        return obj;
    },
    create(base) {
        return AppPRSetReviewDecisionRequest.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseAppPRSetReviewDecisionRequest();
        message.appPublishRequestId = object.appPublishRequestId ?? "";
        message.decision = object.decision ?? 0;
        return message;
    },
};
messageTypeRegistry.set(AppPRSetReviewDecisionRequest.$type, AppPRSetReviewDecisionRequest);
function toTimestamp(date) {
    const seconds = Math.trunc(date.getTime() / 1000);
    const nanos = (date.getTime() % 1000) * 1000000;
    return { seconds, nanos };
}
function fromTimestamp(t) {
    let millis = (t.seconds || 0) * 1000;
    millis += (t.nanos || 0) / 1000000;
    return new globalThis.Date(millis);
}
function fromJsonTimestamp(o) {
    if (o instanceof globalThis.Date) {
        return o;
    }
    else if (typeof o === "string") {
        return new globalThis.Date(o);
    }
    else {
        return fromTimestamp(Timestamp.fromJSON(o));
    }
}
function isSet(value) {
    return value !== null && value !== undefined;
}
