// @devvit/protos/types/devvit/payments/v1alpha/common.js
/**
 * #common.ts
 *
 * @packageDocumentation
 */
/* eslint-disable */
/** TODO: we may want to pull this directly from the pay protos via snootobuf */
export var Environment;
(function (Environment) {
    Environment[Environment["ENVIRONMENT_UNSPECIFIED"] = 0] = "ENVIRONMENT_UNSPECIFIED";
    Environment[Environment["ENVIRONMENT_SANDBOX"] = 1] = "ENVIRONMENT_SANDBOX";
    Environment[Environment["ENVIRONMENT_PRODUCTION"] = 2] = "ENVIRONMENT_PRODUCTION";
    Environment[Environment["UNRECOGNIZED"] = -1] = "UNRECOGNIZED";
})(Environment || (Environment = {}));
export function environmentFromJSON(object) {
    switch (object) {
        case 0:
        case "ENVIRONMENT_UNSPECIFIED":
            return Environment.ENVIRONMENT_UNSPECIFIED;
        case 1:
        case "ENVIRONMENT_SANDBOX":
            return Environment.ENVIRONMENT_SANDBOX;
        case 2:
        case "ENVIRONMENT_PRODUCTION":
            return Environment.ENVIRONMENT_PRODUCTION;
        case -1:
        case "UNRECOGNIZED":
        default:
            return Environment.UNRECOGNIZED;
    }
}
export function environmentToJSON(object) {
    switch (object) {
        case Environment.ENVIRONMENT_UNSPECIFIED:
            return 0;
        case Environment.ENVIRONMENT_SANDBOX:
            return 1;
        case Environment.ENVIRONMENT_PRODUCTION:
            return 2;
        case Environment.UNRECOGNIZED:
        default:
            return -1;
    }
}
