// @devvit/protos/types/devvit/actor/hello/hello.twirp.js
import { TwirpServer, TwirpError, TwirpErrorCode, TwirpContentType, chainInterceptors, } from 'twirp-ts';
import { PingMessage } from './hello.js';
export var HelloMethod;
(function (HelloMethod) {
    HelloMethod["Ping"] = "Ping";
})(HelloMethod || (HelloMethod = {}));
export const HelloMethodList = [HelloMethod.Ping];
export function createHelloServer(service) {
    return new TwirpServer({
        service,
        packageName: 'devvit.actor.hello',
        serviceName: 'Hello',
        methodList: HelloMethodList,
        matchRoute: matchHelloRoute,
    });
}
function matchHelloRoute(method, events) {
    switch (method) {
        case 'Ping':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'Ping' };
                await events.onMatch(ctx);
                return handleHelloPingRequest(ctx, service, data, interceptors);
            };
        default:
            events.onNotFound();
            const msg = `no handler found`;
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleHelloPingRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleHelloPingJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleHelloPingProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
async function handleHelloPingJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = PingMessage.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Ping(ctx, inputReq);
        });
    }
    else {
        response = await service.Ping(ctx, request);
    }
    return JSON.stringify(PingMessage.toJSON(response));
}
async function handleHelloPingProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = PingMessage.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Ping(ctx, inputReq);
        });
    }
    else {
        response = await service.Ping(ctx, request);
    }
    return Buffer.from(PingMessage.encode(response).finish());
}
