// @devvit/protos/types/devvit/actor/reddit/post_filter.twirp.js
import { TwirpServer, TwirpError, TwirpErrorCode, TwirpContentType, chainInterceptors, } from 'twirp-ts';
import { PostFilterRequest, Response } from './post_filter.js';
export var PostFilterMethod;
(function (PostFilterMethod) {
    PostFilterMethod["Filter"] = "Filter";
})(PostFilterMethod || (PostFilterMethod = {}));
export const PostFilterMethodList = [PostFilterMethod.Filter];
export function createPostFilterServer(service) {
    return new TwirpServer({
        service,
        packageName: 'devvit.actor.reddit',
        serviceName: 'PostFilter',
        methodList: PostFilterMethodList,
        matchRoute: matchPostFilterRoute,
    });
}
function matchPostFilterRoute(method, events) {
    switch (method) {
        case 'Filter':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'Filter' };
                await events.onMatch(ctx);
                return handlePostFilterFilterRequest(ctx, service, data, interceptors);
            };
        default:
            events.onNotFound();
            const msg = `no handler found`;
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handlePostFilterFilterRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handlePostFilterFilterJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handlePostFilterFilterProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
async function handlePostFilterFilterJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = PostFilterRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Filter(ctx, inputReq);
        });
    }
    else {
        response = await service.Filter(ctx, request);
    }
    return JSON.stringify(Response.toJSON(response));
}
async function handlePostFilterFilterProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = PostFilterRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Filter(ctx, inputReq);
        });
    }
    else {
        response = await service.Filter(ctx, request);
    }
    return Buffer.from(Response.encode(response).finish());
}
